java-frontend/src/main/java/org/sonar/java/JavaFrontend.java

```diff
@@ -0,0 +1,660 @@
+        return false;
+      }
+
+      IndirectTaintSource other = (IndirectTaintSource) obj;
+      return sources.equals(other.sources);
+    }
+
+  }
+
+  private static class TaintSourceFactory {
+
+    private Map<TaintSource, TaintSource> singletons = new HashMap<>();
+
+    public TaintSource taintFree() {
+      return singletonFor(new IndirectTaintSource());
+    }
+
+    public TaintSource taintedIif(String fqn, List<TaintSource> arguments) {
+      return singletonFor(new DirectTaintSource(fqn, arguments));
+    }
+
+    private TaintSource singletonFor(TaintSource taintSource) {
+      return singletons.computeIfAbsent(taintSource, ts -> ts);
+    }
+
+  }
+
+  private static class TaintProgramState {
+
+    private final ScannedFile src;
+    private final TaintSourceFactory tsf;
+    private final Map<Symbol, TaintSource> taintSources = new HashMap<>();
+    private final Set<TaintSource> callees = new HashSet<>();
+
+    public TaintProgramState(ScannedFile src, TaintSourceFactory tsf) {
+      this.src = src;
+      this.tsf = tsf;
+    }
+
+    private TaintProgramState(TaintProgramState other) {
+      this.src = other.src;
+      this.tsf = other.tsf;
+      this.taintSources.putAll(other.taintSources);
+      this.callees.addAll(other.callees);
+    }
+
+    public TaintProgramState dup() {
+      return new TaintProgramState(this);
+    }
+
+    public SemanticModel semantic() {
+      return src.semantic();
+    }
+
+    private static String fullyQualify(Symbol s) {
+      if (s.isVariableSymbol()) {
+        Symbol owner = s.owner();
+        if (owner.isTypeSymbol()) {
+          // field
+          return ((JavaSymbol.TypeJavaSymbol)owner).getFullyQualifiedName() + "#" + s.name();
+        } else if (owner.isMethodSymbol()) {
+          // parameter
+          JavaSymbol.MethodJavaSymbol m = (JavaSymbol.MethodJavaSymbol)s.owner();
+          int paramId = m.getParameters().scopeSymbols().indexOf(s);
+          if (paramId == -1) {
+            throw new IllegalStateException();
+          }
+          return ((JavaSymbol.MethodJavaSymbol)s.owner()).completeSignature() + "#" + paramId;
+        } else {
+          throw new IllegalArgumentException();
+        }
+      } else if (s.isMethodSymbol()) {
+        return ((JavaSymbol.MethodJavaSymbol)s).completeSignature();
+      } else {
+        throw new IllegalArgumentException();
+      }
+    }
+
+    public static boolean isString(Type type) {
+      return type.is("java.lang.String");
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol) {
+      return taintSourceFor(symbol, Collections.emptyList());
+    }
+
+    public TaintSource taintSourceFor(Symbol symbol, List<TaintSource> arguments) {
+      if (symbol.isMethodSymbol()) {
+        MethodSymbol method = (MethodSymbol)symbol;
+        TaintSource methodTaintSource = tsf.taintedIif(fullyQualify(symbol), arguments);
+
+        if (arguments.stream().anyMatch(ts -> ts.canBeTainted())) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this lambda with a method reference. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1612)